### PR TITLE
ZBUG-563 Fixed sorting by read/unread flag

### PR DESCRIPTION
**Pull request by rupalid**
Add code to fix sorting based on read/unread flag.
Added code to set sortby to date when query contain is:unread
Added code to ignore sort val when recd for sortby readAsc and readDesc
Added unit test

Verified that sort results are correctly fetched when sorting  by unread and all using the UI
All unit test pass

Tests | Failures | Errors | Skipped | Success rate | Time
1327  | 0           | 0         | 10           | 100.00% | 339.735

### CODE REVIEW

File: store/src/java/com/zimbra/cs/index/MessageHit.java [hunk 1]
```diff
@@ -178,4 +179,44 @@ public ConversationHit getConversationResult() throws ServiceException {
         }
         return conversationHit;
     }
+    
+    static final int compareByReadFlag(boolean ascending, ZimbraHit lhs, ZimbraHit rhs) {
```
[the remark this answers is not in the document]
Comment (rupalid) [reply store/src/java/com/zimbra/cs/index/MessageHit.java]: Fixed

File: store/src/java/com/zimbra/cs/index/MessageHit.java [hunk 2]
```diff
@@ -178,4 +179,44 @@ public ConversationHit getConversationResult() throws ServiceException {
...
+        else
+            return retVal;
+    }
+
+    /**
+     * @param lhs
+     * @return
+     * @throws ServiceException 
+     */
+    private static int getReadStatus(ZimbraHit zh) throws ServiceException {
```
[the remark this answers is not in the document]
Comment (rupalid) [reply store/src/java/com/zimbra/cs/index/MessageHit.java]: Fixed

File: store/src/java/com/zimbra/cs/index/ConversationHit.java [hunk 3]
```diff
@@ -129,4 +131,41 @@ public Conversation getConversation() throws ServiceException {
...
+     * @param lhs
+     * @return
+     * @throws ServiceException 
+     */
+    private static int getReadStatus(ZimbraHit zh) throws ServiceException {
+        if (zh instanceof ProxiedHit) {
+            return ((ProxiedHit) zh).getElement().getAttributeInt(MailConstants.A_UNREAD);
+        }
+        else {
+            boolean unread = ((ConversationHit) zh).getMailItem().isUnread();
```
[the remark this answers is not in the document]
Comment (rupalid) [reply store/src/java/com/zimbra/cs/index/ConversationHit.java]: Removed the cast

File: store/src/java/com/zimbra/cs/index/ConversationHit.java [hunk 4]
```diff
@@ -129,4 +131,41 @@ public Conversation getConversation() throws ServiceException {
...
+            long right = getReadStatus(rhs);
+            long result = right - left;
+            if (result > 0)
+                retVal = 1;
+            else if (result < 0)
+                retVal = -1;
+            else
+                retVal = 0;
+        } catch (ServiceException e) {
+            ZimbraLog.index.info("Caught ServiceException trying to compare ConvHit %s to ConvHit %s",
```
[the remark this answers is not in the document]
Comment (rupalid) [reply store/src/java/com/zimbra/cs/index/ConversationHit.java]: Fixed

File: store/src/java-test/com/zimbra/cs/index/SearchParamsTest.java [hunk 5]
```diff
@@ -35,5 +35,12 @@ public void parseLocale() {
         Assert.assertEquals(new Locale("en"), SearchParams.parseLocale("en"));
         Assert.assertEquals(new Locale("en", "US", "MAC"), SearchParams.parseLocale("en_US-MAC"));
     }
+    
+    
+    public void testIsSortByReadFlag() {
+        Assert.assertFalse(SearchParams.isSortByReadFlag(SortBy.DATE_ASC));
```
[the remark this answers is not in the document]
Comment (rupalid) [reply store/src/java-test/com/zimbra/cs/index/SearchParamsTest.java]: Fixed

File: store/src/java/com/zimbra/cs/index/ConversationHit.java [hunk 6]
```diff
@@ -129,4 +131,41 @@ public Conversation getConversation() throws ServiceException {
...
+            long left = getReadStatus(lhs);
+            long right = getReadStatus(rhs);
+            long result = right - left;
+            if (result > 0)
+                retVal = 1;
+            else if (result < 0)
+                retVal = -1;
+            else
+                retVal = 0;
+        } catch (ServiceException e) {
```
[the remark this answers is not in the document]
Comment (rupalid) [reply store/src/java/com/zimbra/cs/index/ConversationHit.java]: Fixed

File: store/src/java/com/zimbra/cs/index/ReSortingQueryResults.java [hunk 7]
```diff
@@ -140,6 +150,49 @@ private void bufferAllHits() throws ServiceException {
...
+                            return MessageHit.compareByReadFlag(true, lhs, rhs);
+                        }
+                    }
+                };
+                break;
+            case READ_DESC:
+                comp = new Comparator<ZimbraHit>() {
+                    @Override
+                    public int compare(ZimbraHit lhs, ZimbraHit rhs) {
+                        if (lhs  instanceof ConversationHit) {
```
Comment (shrikant-prasad) [inline store/src/java/com/zimbra/cs/index/ReSortingQueryResults.java]: no need to check the type as in all three cases same method is being called.

File: store/src/java/com/zimbra/cs/index/ReSortingQueryResults.java [hunk 8]
```diff
@@ -140,6 +150,49 @@ private void bufferAllHits() throws ServiceException {
...
+                            return MessageHit.compareByReadFlag(true, lhs, rhs);
+                        } 
+                    }
+                };
+                break;
+            case READ_ASC:
+                comp = new Comparator<ZimbraHit>() {
+                    @Override
+                    public int compare(ZimbraHit lhs, ZimbraHit rhs) {
+                        if (lhs  instanceof ConversationHit) {
```
Comment (shrikant-prasad) [inline store/src/java/com/zimbra/cs/index/ReSortingQueryResults.java]: no need to check the type as in all three cases same method is being called.

File: store/src/java/com/zimbra/cs/index/ReSortingQueryResults.java [hunk 9]
```diff
@@ -140,6 +150,49 @@ private void bufferAllHits() throws ServiceException {
         Comparator<ZimbraHit> comp;
         switch (sort) {
             default:
+                comp = new Comparator<ZimbraHit>() {
+                    @Override
+                    public int compare(ZimbraHit lhs, ZimbraHit rhs) {
+                        if (lhs  instanceof TaskHit) {
+                            return TaskHit.compareByDueDate(true, lhs, rhs);
+                        } else if (lhs  instanceof ConversationHit) {
```
Comment (shrikant-prasad) [inline store/src/java/com/zimbra/cs/index/ReSortingQueryResults.java]: can combine the last three else into one.

**Pull request merged**